Position Control Refactoring Part 6

MaEtUgR opened the pull request:
**Describe problem solved by this pull request**
This is part 6 of the refactoring from the monster pr #12072 and should be easy to understand and review since the diff is manageable.

**Describe your solution**
- Replace velocity PID calculations for each axis separate spread across the whole function with a one-line `Vector3f` formula. https://github.com/PX4/Firmware/commit/af3ede01353fdb80a4329ab9a91536a46d4ae69a
- Use `addIfNotNan` helper functions for feed-forward summation with potentially NAN values in each axis. This simplifies velocity feed-forward and enables unit thrust feed-forward (acceleration to come in next pr).
- Make sure the hacky way horizontal thrust setpoints get combined with altitude and vertical velocity setpoints in "Altitude" mode still works to not break any functionality. This part can be removed soon with acceleration setpoints.
- Refactor horizontal thrust limit calculation.
- Replace velocity integral summation for each axis separate spread across the whole function with a one-line `Vector3f` formula. Do anti-windup by reducing the summed up `vel_error` before that.
- Make sure NAN yaw/yawrate setpoints don't get into the attitude setpoint because the attitude controller doesn't support this yet.
- Replace all the checks in the `interfaceMapping()` and externally in the `mc_pos_control` module which handled every single case separately, were called in the setter, introduced internal setpoint changes that had to be reverted before logging. Instead, use a set of simple rules in the method `PositionControl._updateSuccessful()` that are all checked and returned with `PositionControl.update()`.
- Remove obsolete `skip_controller` case which was used when all the thrust setpoints were already set in the input. Since thrust feed-forward is supported this case is implicitly supported.

**Test data / coverage**
SITL hover goto, and manual position altitude tested and unit tests pass.

**Additional context**
Part 5 that goes before this: #13701

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControl.cpp:
```diff
@@ -317,12 +317,11 @@ void PositionControl::_velocityControl(const float dt)
-		float mag = thrust_sp_xy.length();
-		_thr_sp(0) = _thr_sp(0) / mag * thrust_max_NE;
-		_thr_sp(1) = _thr_sp(1) / mag * thrust_max_NE;
+	if (thrust_sp_xy_norm > thrust_max_NE) {
```
use 
```
    bool longerThan(Type testVal) const {
        return norm_squared() > testVal*testVal;
    }
```
from the matrix library

MaEtUgR replied:
I thought about that but the norm anyways needs to be calculated... IMO it doesn't help readability and is less efficient but if it helps in your eyes we can put:
```
const Vector2f thrust_sp_xy(_thr_sp);

if (_thr_sp.xy().longerThan(thrust_max_NE)) {
	_thr_sp.xy() = Vector2f(_thr_sp).normalized() * thrust_max_NE;
}
```

MaEtUgR replied:
Like we just discussed it doesn't help in that particular case.

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControlTest.cpp:
```diff
@@ -138,7 +144,7 @@ TEST_F(PositionControlBasicDirectionTest, PositionDirection)
 	_input_setpoint.y = .1f;
 	_input_setpoint.z = -.1f;
-	runController();
+	EXPECT_EQ(runController(), true);
```
```suggestion
	EXPECT_TRUE(runController());
```

MaEtUgR replied:
🤦‍♂

bresch commented on src/modules/mc_pos_control/PositionControl/PositionControlTest.cpp:
```diff
@@ -246,3 +252,78 @@ TEST_F(PositionControlBasicTest, InputCombinationsPositionVelocity)
+
+TEST_F(PositionControlBasicTest, SetpointValiditySimple)
+{
+	EXPECT_EQ(runController(), false);
```
```suggestion
	EXPECT_FALSE(runController());
```

MaEtUgR commented on src/modules/mc_pos_control/PositionControl/ControlMath.cpp:
```diff
@@ -215,4 +215,32 @@ bool cross_sphere_line(const Vector3f &sphere_c, const float sphere_r,
+	// Addition is NAN or both are NAN, nothing to do
+}
+
+void addIfNotNanVector(Vector3f &setpoint, const Vector3f &addition)
```
More clear, yes.

MaEtUgR commented:
I
- rebased cleanly
- addressed the review, thanks @bresch 
- removed obsolete reassignments in `mc_pos_control` for logging which were only necessary before this pr

LorenzMeier commented:
@dagar Coverage is going down as it looks like the old flags are not running here - any idea why that is?

MaEtUgR commented:
I wouldn't know why coverage should decrease. I increased unit testing by covering validity checks for all possible combinations of inputs: https://github.com/PX4/Firmware/pull/14017/files#diff-8453f440bea376d8b827785273aa4848R256-R329 And the position control directions and limits tests I already added in earlier prs of this refactoring.
The report doesn't seem to be representative e.g. ControlMath has 100% unit test coverage and that's where it reports `-35.7%` 🤔 .

dagar commented:
Jenkins Firmware-SITL_tests was only configured to run on a subset of branches.
 - master/beta/stable
 - pr-*

I think that can be lifted, although it could also be nice if we had some simple branch naming conventions.

MaEtUgR commented:
I did some outdoor flight testing with this branch:
Carefully started in Manual such that I can easily fall back and successfully tested altitude and position mode:
https://logs.px4.io/plot_app?log=6a289be6-bca0-4d75-b458-c9c79e6962c3
Takeoff in position, some testing, land in altitude:
https://logs.px4.io/plot_app?log=51559d82-7e7b-418e-986d-84762243dcf6
Takeoff in altitude, some testing, land in position:
https://logs.px4.io/plot_app?log=8b7c5d63-9fff-4be2-adc1-bc94669c79d1
It flew flawlessly, I didn't find any issues.

MaEtUgR commented:
@PX4/testflights It would be nice if you could verify more generally also in mission to be sure. I feel a lot more confident now but in the worst case Manual mode would be the untouched fallback.

Junkim3DR commented:
### Tested on NXP FMUK66 v3

**Modes Tested**
- Altitude Mode: Good.

**Procedure**
Arm and Take off in altitude mode and test only altitude.

**Notes**
No issues noted, good flight in general.

**Log**
- https://review.px4.io/plot_app?log=f954c612-5ebc-451a-9cf3-1a33730200f7

jorge789 commented:
### Tested on CUAV nano V5

**Modes Tested**

Altitude Mode: Good.
Procedure
Arm and Take off in altitude mode and test only altitude.

**Notes**
No issues noted, good flight in general.

**Log:**  https://review.px4.io/plot_app?log=2417cd21-20c3-4e84-b08e-fb0b627301b4

dannyfpv commented:
Tested on pixhawk4 v5 f-450
Modes Tested

Position Mode: Good.

Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.
Procedure
Arm and Take off in position mode,  to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL.

Notes
No issues noted, good flight in general.

Log


https://review.px4.io/plot_app?log=8217e81b-7d3a-4e09-bb2d-ffd3791ae784

jorge789 commented:
### Tested on CUAV nano V5 

**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.
**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint then trigger RTL.

**Notes**
No issues noted, good flight in general.

**Log:**  https://review.px4.io/plot_app?log=03a6d7a6-2a9b-4511-8de4-580f888f8781

MaEtUgR commented:
According to @dagar 's feedback I got rid of the "address review comments" commit squashing the changes into the respective original commits. I made sure the diff of all the commits in this pr stay exactly identical. Then I rebased on master so it's ready to merge.
